soroban-sdk/src/bytes.rs

```diff
@@ -602,6 +605,70 @@ impl Bytes {
     pub fn iter(&self) -> BytesIter {
         self.clone().into_iter()
     }
+
+    /// Copy the bytes into a buffer of given size.
+    ///
+    /// Returns the buffer and a range of where the bytes live in the given
+    /// buffer.
+    ///
+    /// Suitable when the size of the bytes isn't a fixed size but it is known
+    /// to be under a certain size, or failure due to overflow is acceptable.
+    ///
+    /// ### Panics
+    ///
+    /// If the size of the bytes is larger than the size of the buffer. To avoid
+    /// this, first slice the bytes into a smaller size then convert to a
+    /// buffer.
+    #[must_use]
+    pub fn to_buffer<const B: usize>(&self) -> BytesBuffer<B> {
+        let mut buffer = [0u8; B];
+        let len = self.len() as usize;
+        {
+            let slice = &mut buffer[0..len];
+            self.copy_into_slice(slice);
+        }
+        BytesBuffer {
+            buffer,
+            range: 0..len,
+        }
+    }
+
+    /// Copy the bytes into a Rust alloc Vec of size matching the bytes.
+    ///
+    /// Returns the Vec. Allocates using the built-in allocator.
+    ///
+    /// Suitable when the size of the bytes isn't a fixed size and the allocator
+    /// functionality of the sdk is enabled.
+    #[cfg(feature = "alloc")]
+    #[must_use]
+    pub fn to_alloc_vec(&self) -> alloc::vec::Vec<u8> {
+        let len = self.len() as usize;
+        let mut vec = alloc::vec::from_elem(0u8, len);
+        self.copy_into_slice(&mut vec);
+        vec
+    }
+}
+
+/// A BytesBuffer stores bytes in a buffer of given size and makes the bytes
+/// stored available.
```

Comment:
```suggestion
/// A `BytesBuffer` stores a variable number of bytes, up to a fixed limit `B`, in a fixed-size
/// non-heap-allocated structure. It is a minimal wrapper around a fixed-size `[u8;B]` byte
/// array and a length field indicating the amount of the byte array containing meaningful data.
```

Reply:
Given this type really only exists to cope with awkwardness around heap-allocation and size-variability, I think it'd be helpful to highlight exactly what it means in those terms in its doc-comment.